### PR TITLE
fixes for manual reporting

### PR DESCRIPTION
**Pull request by sumeet**
travis is broken because of an integration spec, disregard

#### fix 1: fix issues reporting with an error string instead of an exception:
- rollbar will now send the backtrace
- rollbar will print the error class and error message properly. before
  this change, it would not include the error class

proof

### CODE REVIEW

File: lib/exceptions/backends/rollbar.rb [hunk 1]
```diff
@@ -42,14 +42,32 @@ def clear_context
...
+      # if reporting via the `error_class` and `error_message` args, we need to
+      # trick the rollbar reporter into thinking it's an actual exception. by
+      # doing that, rollbar will report the error class and message correctly,
+      # and show a stack trace, which it doesn't do when passing it a string.
+      def maybe_turn_into_exception(error_class_or_exception_or_string,
+                                    error_message)
+        if error_class_or_exception_or_string.is_a?(Exception)
+          error_class_or_exception_or_string
+        else
+          Class.new(StandardError) do
```
Comment (v-yarotsky) [inline lib/exceptions/backends/rollbar.rb]: hopefully these are garbage collected :)
(I believe they are)

File: lib/rack/exceptions.rb [hunk 2]
```diff
@@ -1,26 +1,36 @@
 module Rack
   class Exceptions
+    # XXX: applications are typically configured with a single global instance
```
Comment (v-yarotsky) [inline lib/rack/exceptions.rb]: Looks legit, but why the `XXX:`?

File: lib/exceptions/backends/rollbar.rb (hunk 1 above)
Comment (sumeet) [reply lib/exceptions/backends/rollbar.rb]: lol, good thing rollbar is accepting of these fakes. from my research, the only legit way of giving a class a name is by using `const_set`, which would've been crazy

File: lib/rack/exceptions.rb (hunk 2 above)
Comment (sumeet) [reply lib/rack/exceptions.rb]: i felt gross doing this, i thought there could've been a way around it by passing references the NORMAL way. but if i think about it more, i think it's a necessary evil b/c we want Exceptions to be global. removing the XXX haha

File: lib/exceptions/backends/rollbar.rb (hunk 1 above)
Comment (hamiltop) [reply lib/exceptions/backends/rollbar.rb]: So... any chance these are not GC'd? We saw some OOMs start happening in r101-api when it was updated.

Comment (v-yarotsky) [reply lib/exceptions/backends/rollbar.rb]:
```ruby
puts "pre"
puts "classes: #{ObjectSpace.count_objects[:T_CLASS]}"
puts "objects: #{ObjectSpace.count_objects[:T_OBJECT]}"
1_000_000.times do |i|
  Class.new(StandardError) do
    define_singleton_method(:name) do
      "name_#{i}"
    end

    define_method(:message) do
      "message_#{i}"
    end
  end.new
end
puts "post"
puts "classes: #{ObjectSpace.count_objects[:T_CLASS]}"
puts "objects: #{ObjectSpace.count_objects[:T_OBJECT]}"
GC.start
puts "post gc"
puts "classes: #{ObjectSpace.count_objects[:T_CLASS]}"
puts "objects: #{ObjectSpace.count_objects[:T_OBJECT]}"
```

```
pre
classes: 550
objects: 1687
post
classes: 3501
objects: 2576
post gc
classes: 552
objects: 1684
```

TL;DR they are

**Comment by sumeet** [comment]
damn.....

though it's perhaps possible some code in our stack is preventing this guy
from getting deallocated. i will try to repro locally

2017년 9월 22일 (금) 오전 2:43, Vlad Yarotsky <notifications@github.com>님이 작성:
